### PR TITLE
feat(storage-resize-images): Adds support for absolute paths restrictions for Storage Resize Image

### LINKED ISSUE #426: Absolute paths restrictions for Storage Resize Image
### [READ] Step 1: Are you in the right place?

Issues filed here should be about a feature request for a **specific extension in this repository**. To file a feature request that affects multiple extensions or the Firebase Extensions platform, please reach out to
[Firebase support](https://firebase.google.com/support/troubleshooter/report/features/) directly.

### [REQUIRED] Step 2: Extension name

This feature request is for extension: storage-resize-images

### What feature would you like to see?

It would be great to have the possibility to define a list of absolute paths inside the bucket in which listen for new image uploads to be resized.

The absolute paths to be provided in the configuration option could be a comma-separated list, like this:

`/users/avatars,/posts/pictures,/design/resources/icons`

This will mean that any image uploaded inside these paths, or inside any folder at any depth inside these will be resized. 
And of course, any image outside of these paths will not be resized.

### How would you use it?

This will allow us to limit the images to be resized by the extension. For instance, the user's avatars could be resized, but what about if we don't want resized images for posts pictures? 

This feature could be very helpful to save much more space in the storage and not waste the resource of Firebase.

### PR DESCRIPTION
**Pull request by yulioaj290**
This adds support to define a list of absolute paths inside the bucket in which listen for new image uploads to be resized.

The absolute paths to be provided in the configuration option is a comma-separated list, like this:

`/users/avatars,/posts/pictures,/design/resources/icons`

This means that any image uploaded inside these paths, or inside any folder at any depth inside these will be resized.
And of course, any image outside of these paths will not be resized.

Fixes #426

### CODE REVIEW

**Comment by yulioaj290** [comment]
@googlebot I signed it!

**Comment by Ehesp** [comment]
Thanks for this PR! We'll get it reviewed and tested out.

File: storage-resize-images/functions/src/index.ts [hunk 1]
```diff
@@ -59,6 +59,8 @@ export const generateResizedImage = functions.storage.object().onFinalize(
   async (object): Promise<void> => {
     logs.start();
     const { contentType } = object; // This is the image MIME type
+    const absolutePathList = config.absolutePathList.split(","); // Convert list to an array
+    const tmpFilePath = path.dirname(object.name);
```
Comment (Ehesp) [inline storage-resize-images/functions/src/index.ts:63]: `dirname` does not return a leading `/` if the provided path doesn't start with one. The regex for `ABSOLUTE_PATH_LIST` requires one, therefore this will never match. There needs to be a leading `/`.

**Review by Ehesp — CHANGES_REQUESTED** [review]

**Comment by Ehesp** [comment]
Thanks for the PR - there are some changes needed for this to work. Also, could you run the `build` script from the `package.json` to regen the source files.

**Comment by yulioaj290** [comment]
@Ehesp I already followed your recommendations. Let me know if I need to do something else :)

**Review by Ehesp — APPROVED** [review]
LGTM

**Comment by yulioaj290** [comment]
@Ehesp I updated the PR with the changes from `upstream`. Tell me if should I do anything else :smile:

**Comment by jhuleatt** [comment]
Hi @yulioaj290! Thanks for the feature request. Given the huge array of configuration options that sharp provides, and our desire to keep the parameters for this extension limited to a small set of essential config, we're not able to add this feature.

However, we have published Cloud Functions sample code [here](https://github.com/firebase/functions-samples/tree/master/image-sharp) that should be easy to customize to your exact needs. I hope that helps with your use case.

File: storage-resize-images/extension.yaml [hunk 2]
```diff
@@ -146,6 +146,21 @@ params:
...
     required: false
 
+  - param: ABSOLUTE_PATH_LIST
+    label: List of absolute paths for resized images
+    description: >
+      A comma-separated list of absolute paths to restrict the 
+      images to be resized. For instance, if you want to resize only 
+      the images stored in `/users/pictures`, then other images uploaded 
+      outside that specific path will not be resized. If you prefer to 
+      resize whole the images uploaded to the bucket, leave this field empty.
```
Comment (russellwheatley) [inline storage-resize-images/extension.yaml:156]: I think reword this part of the description:
"For instance, if you want to resize only the images stored in `/users/pictures`, then other images uploaded outside that specific path will not be resized. If you prefer to resize whole the images uploaded to the bucket, leave this field empty."

To something like this:
"For example, to only resize the images stored in `/users/pictures` directory, specify the path `/users/pictures`. If you prefer to resize every image uploaded to the storage bucket, leave this field empty."

Comment (russellwheatley) [reply storage-resize-images/extension.yaml:156]: Hey @kevinthecheung, do you mind taking a look at this and make sure it sounds ok, please? Thanks.

**Review by russellwheatley — CHANGES_REQUESTED** [review]
Thanks for your work, @yulioaj290! I think if you change the wording in the description and sort out the merge conflicts this PR should be good to go. 

@jhuleatt This PR ought to be backwards compatible as this [line](https://github.com/firebase/extensions/pull/427/files#diff-5e32c49ddbaed10fcb44f088672f2fa9R62) takes care of any extension without the latest extension parameter: `absolutePathList`.

**Comment by yulioaj290** [comment]
@Ehesp @jhuleatt @russellwheatley 
I added another config option to this PR: `EXCLUDE_PATH_LIST`

Please review again the PR for your approval and let me know if any change is needed.

File: storage-resize-images/extension.yaml (hunk 2 above)
Comment (yulioaj290) [reply storage-resize-images/extension.yaml]: @russellwheatley It's done! :+1:

**Review by dackers86 — APPROVED** [review]

**Comment by dackers86** [comment]
Test Cases Ran

Initial Config

- [x] Uploads and resizes at root
- [x] Uploads and resizes in a sub folder

Allowed Paths

- [x] Does not resize ay the root
- [x] Resizes in the specified paths
- [x] Does not resize in an unspecific path

Disallowed Paths

- [x] Resizes at the root
- [x] Does not resize in specified paths
- [x] Resizes in other paths

Allowed & Disallowed Paths

- [x] Does not resize at the root
- [x] Resizes in allowed paths
- [x] Does not resize in disallowed paths

Deletes Original File
- [x] Deletes original file if checked


Resizing
- [x] Successfully resizes based on resize values

File: storage-resize-images/extension.yaml [hunk 3]
```diff
@@ -168,6 +168,36 @@ params:
...
+  - param: ABSOLUTE_PATH_LIST
+    label: List of absolute paths for resized images
+    description: >
+      A comma-separated list of absolute paths to restrict the 
+      images to be resized. For example, to only resize the images 
+      stored in `/users/pictures` directory, specify the path `/users/pictures`. 
+      If you prefer to resize every image uploaded to the storage bucket, 
+      leave this field empty.
+    type: string
+    example: "/users/avatars,/design/pictures"
```
Comment (jhuleatt) [inline storage-resize-images/extension.yaml:180]: Let's modify this param and its label/description so that its purpose is more clear. Here's my attempt:

```suggestion
  - param: INCLUDE_PATH_LIST
    label: Specific paths that include images to resize
    description: >
      Restrict storage-resize-images to only resize specific areas of your Storage bucket by 
      supplying a comma-separated list of absolute paths. For example, to only resize the images 
      stored in `/users/pictures` directory, specify the path `/users/pictures`. 
      If you prefer to resize every image uploaded to the storage bucket, 
      leave this field empty.
    type: string
    example: "/users/avatars,/design/pictures"
```

@kevinthecheung any thoughts on how to phrase this?

Comment (jhuleatt) [reply storage-resize-images/extension.yaml:180]: Once we decide on wording, let's check if the same edits need to be applied to `EXCLUDE_PATH_LIST`

Comment (jhuleatt) [reply storage-resize-images/extension.yaml:180]: We should also note somewhere that, no matter what paths are set here, the extension will still trigger for every upload to a bucket. It will just exit itself if it notices that it was triggered on a file it isn't supposed to resize

Comment (yulioaj290) [reply storage-resize-images/extension.yaml:180]: I like your suggestion about to reword the option `ABSOLUTE_PATH_LIST` to `INCLUDE_PATH_LIST`, it sounds better aligned to the sense of the config option itself. So it will be `include / exclude` :+1:

Comment (kevinthecheung) [reply storage-resize-images/extension.yaml:180]: How about:
```
 - param: INCLUDE_PATH_LIST
    label: Paths that contain images you want to resize
    description: >
      Restrict storage-resize-images to only resize images in specific locations in your Storage bucket by 
      supplying a comma-separated list of absolute paths. For example, to only resize the images 
      stored in `/users/pictures` directory, specify the path `/users/pictures`. 
      If you prefer to resize every image uploaded to the storage bucket, 
      leave this field empty.
    type: string
    example: "/users/avatars,/design/pictures"
```

**Comment by russellwheatley** [comment]
@googlebot I consent.

**Comment by russellwheatley** [comment]
@jhuleatt could you manually confirm cla consent for me, please? 😄

**Comment by yulioaj290** [comment]
@googlebot I consent.

**Comment by russellwheatley** [comment]
@Ehesp Could you give your consent, please? I can't see it and I noticed you've added commits.

File: storage-resize-images/extension.yaml (hunk 3 above)
Comment (jhuleatt) [reply storage-resize-images/extension.yaml]: Looks like Kevin's suggestion has been added. Closing out this thread.

**Comment by jhuleatt** [comment]
> A Googler has manually verified that the CLAs look good.
> 
> **(Googler, please make sure the reason for overriding the CLA status is clearly documented in these comments.)**

Manual override because CLA bot was confused by commits from @russellwheatley and @Ehesp, even though they've signed the CLA.

**Comment by russellwheatley** [comment]
I updated the logic for testing paths as there were edge cases that weren't captured. I tested the implementation again.

Here are function logs using the `INCLUDE_PATH_LIST=['test/picture']`:

This screenshot shows the image resize being blocked as it was inserted at the root of the bucket:
Screenshot 2020-11-17 at 16 06 06

This screenshot shows the image resize being allowed as it was inserted at `/test/picture`:
Screenshot 2020-11-17 at 16 09 32


Here are function logs for a deployed extension using the `EXCLUDE_PATH_LIST=['deny/picture']`:

This screenshot shows the image resize being blocked as it was inserted at `deny/picture`:
Screenshot 2020-11-17 at 16 13 12


This screenshot shows the image being allowed as it was inserted at `/`:
Screenshot 2020-11-17 at 16 11 38

**Comment by russellwheatley** [comment]
@jhuleatt This was failing code coverage by 0.01% 🙃  so I started testing the main module for `storage-resize-images` extension. It should pass now 🤞

**Review by dackers86 — APPROVED** [review]

**Review by russellwheatley — APPROVED** [review]

**Comment by jhuleatt** [comment]
Looks like the checks are stuck due to [an issue with GitHub Actions](https://www.githubstatus.com/incidents/sxthxvdlljk4).

I ran the tests on Node 10 and Node 12 locally and all passed.
